libethereum/State.cpp

```diff
@@ -535,57 +535,54 @@ pair<TransactionReceipts, bool> State::sync(BlockChain const& _bc, TransactionQu
 					if (req > got)
 					{
 						// too old
-						for (Transaction const& t: m_transactions)
-							if (t.from() == i.second.from())
+						for (Transaction const& mt: m_transactions)
+						{
+							if (mt.from() == t.from())
 							{
-								if (t.nonce() < i.second.nonce())
-								{
-									cnote << i.first << "Dropping old transaction (nonce too low)";
-									_tq.drop(i.first);
-								}
-								else if (t.nonce() == i.second.nonce() && t.gasPrice() <= i.second.gasPrice())
-								{
-									cnote << i.first << "Dropping old transaction (gas price lower)";
-									_tq.drop(i.first);
-								}
+								if (mt.nonce() < t.nonce())
+									cnote << t.sha3() << "Dropping old transaction (nonce too low)";
+								else if (mt.nonce() == t.nonce() && mt.gasPrice() <= t.gasPrice())
+									cnote << t.sha3() << "Dropping old transaction (gas price lower)";
 							}
+						}
+						_tq.drop(t.sha3());
```

Comment:
The intent of the original logic was to keep old transactions, currently in the State, with a higher gas price around, since if we don't mine this block, we will have a chance to reorder the next block with the higher-gas-price transaction and drop the one which is currently up for mining.

If the block is mined, then it may stick around in the queue (until restart, anyway), however it's very much an edge case, and is an acceptably small issue. After refactoring the tx queue, it'll be good to have a cleaning pass after each block which removes these opportunistic transactions (same sender, lower nonce).